TGUI: Autolathe!

Couls opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Moves autolathe over to TGUI

Features include:
- Dropdown Materials list, with colorcoded materials(Bananium has a yellow bar!)
- Categories are now easier to switch through as the autolathe all fits on one scrollable page
- You can now see the most you can make with your materials of any design, rather than just stacks
- Can now print a specific amount

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
more responsive UI
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

https://user-images.githubusercontent.com/30060146/128892263-c5604a3b-9a2d-4c98-824b-5370cbf62589.mp4



## Changelog
:cl:
add: TGUI autolathe
expansion: You can now input a custom amount to print(hard limit at 50 or maximum stack size)
expansion: Can now search and change categories while the autolathe is busy, line up those designs!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Couls committed:
TGUI: Autolathe!

Colovorat commented:
holy fuck, thank you

combustor commented:
god bless you

Urumasi commented:
What if we had https://github.com/BeeStation/BeeStation-Hornet/pull/4210 😳

AMonkeyThatCodes commented:
How did you handle items with custom mats? their UI kind of sucks atm

maxymax13 commented:
what about the other lathey stuff like protolathes techfabs circuit printers n stuff 😳

Couls commented:
> How did you handle items with custom mats? their UI kind of sucks atm

haven't changed it, clicking a toolbox opens up the input screen like normal

maxymax13 commented:
also it's cm³ not cm₃, subscript is for chemistry and superscript is for powers

Couls commented:
> what about the other lathey stuff like protolathes techfabs circuit printers n stuff 😳

if people like this one I might look into the others, should be a tad bit easier

AMonkeyThatCodes commented:
> > How did you handle items with custom mats? their UI kind of sucks atm
> 
> haven't changed it, clicking a toolbox opens up the input screen like normal

:(

stylemistake commented:
This must be the fifth variant of an autolathe UI I've ever seen made from scratch.

Mothblocks commented:
those categories should have `selected` enabled (green) once you click on them

stylemistake commented:
Also js formatting is fucked, did ESLint just die?

KittyNoodle commented:
Fucking finally, current auto-lathes are glitchy and unresponsive. This may be a feature during a feature freeze but we need it so bad, current auto lathe GUI is really bad

Couls committed:
fixes and adjustments

Fix formatting, superscript instead of subscript, selected on category thats currently active, display a message when there are no designs found

Couls commented:
> also it's cm³ not cm₃, subscript is for chemistry and superscript is for powers



> Also js formatting is fucked, did ESLint just die?



> those categories should have `selected` enabled (green) once you click on them

all of the above fixed with the latest commit

Couls committed:
appease the linter

SplinterGP commented:
Would be cool if autolathes had a queue, paradise one does and also has a UI with it

stylemistake commented on code/game/machinery/autolathe.dm:
```diff
@@ -61,104 +60,126 @@
-		if(href_list["make"])
-
+	if(action == "menu")
+		selected_category = null
+		matching_designs.Cut()
+
+	if(action == "category")
+		selected_category = params["selectedCategory"]
+		matching_designs.Cut()
+
+	if(action == "search")
+		matching_designs.Cut()
+
+		for(var/v in stored_research.researched_designs)
+			var/datum/design/D = SSresearch.techweb_design_by_id(v)
+			if(findtext(D.name,params["to_search"]))
+				matching_designs.Add(D)
+	if(action == "make")
+		if (!busy)
 			/////////////////
 			//href protection
-			being_built = stored_research.isDesignResearchedID(href_list["make"])
+			being_built = stored_research.isDesignResearchedID(params["id"])
 			if(!being_built)
 				return
```
You must return TRUE at the end of each ui_act branch, otherwise UI will not immediately update.

stylemistake commented on tgui/packages/tgui/interfaces/Autolathe.js:
```diff
@@ -0,0 +1,200 @@
+                  ) || <Table.Row><Table.Cell>{"No designs found."}</Table.Cell></Table.Row>}
+              </Table>
+            </Flex>
+          </Section>)}
```
Don't stack closing brackets like this, it's bad manners.

stylemistake commented on tgui/packages/tgui/interfaces/Autolathe.js:
```diff
@@ -0,0 +1,200 @@
+                          maxValue={design.maxmult}
+                          disabled={design.buildable}
+                          backgroundColor={design.buildable ? '#999999' : 'default'}
+                          onCommit={(e, value) =>
+                            act('make', {
+                              id: design.id,
+                              multiplier: value,
+                            })} />
```
```suggestion
                          onCommit={(e, value) => act('make', {
                            id: design.id,
                            multiplier: value,
                          })} />
```

stylemistake commented on tgui/packages/tgui/interfaces/Autolathe.js:
```diff
@@ -0,0 +1,200 @@
+              <Button
+                icon="times"
+                content="Close Category"
+                onClick={() =>
+                {
```
This is not C++

```suggestion
                onClick={() => {
```

stylemistake commented on tgui/packages/tgui/interfaces/Autolathe.js:
```diff
@@ -0,0 +1,200 @@
+            placeholder="Search Recipes..."
+            selfClear
+            onChange={(e, value) => {
+              if (value.length)
+              {
```
```suggestion
              if (value.length) {
```

Couls committed:
Apply suggestions from code review

Co-authored-by: Aleksej Komarov <stylemistake@gmail.com>

Couls committed:
more review fixes

moved closing brackets to match rest of js file
added returns to ui_act
changed up ui_data code to check multipliers only if the design is buildable

Couls committed:
some more style/format of the js file
